locator/topology.cc

```diff
@@ -73,8 +79,21 @@ void topology::update_endpoint(const inet_address& ep, endpoint_dc_rack dr)
     tlogger.debug("update_endpoint: {} {}/{}", ep, dr.dc, dr.rack);
     _dc_endpoints[dr.dc].insert(ep);
     _dc_racks[dr.dc][dr.rack].insert(ep);
-    _datacenters.insert(dr.dc);
-    _current_locations[ep] = std::move(dr);
+    auto [datacenters_it, new_dc] = _datacenters.insert(dr.dc);
+    _current_locations[ep] = dr;
+
+    location_metric m = 0;
+    auto [dc_it, dc_emplaced] = _dc_radix.try_emplace(dr.dc, _dc_radix.size());
+    location_metric dc_metric = dc_it->second;
```

Comment:
I think it's better to use just 0/1 (and allocate just one bit for each component). This way, two different datacenters or two different nodes in the same rack are the same distance away and we don't favor one dc over another based on some irrelevant criteria. The code becomes simpler too, because you don't need to assign different nonzero values.

Reply:
Actually, that only works if sort_by_proximity is only called for this node (why wouldn't it?)

In all but one case I could verify that this is the case, probably the last one is too.

Reply:
If we want to preserve that DCs different from the reference node DC are indistinguishable and similarly, rack different than the reference rack in the same DC, we need to compute the metric on each call, which I wanted to avoid, in order to reduce the number of logical branches.

Since the language does guarantee that bool to int conversion will produce 0 or 1 (See https://en.cppreference.com/w/cpp/language/implicit_conversion), we can convert the boolean comparison value of dc/rack to single bit with branching.